HPCC-20583 Allow DESDL application transforms to affect entire request

afishbeck opened the pull request:
Some information can be sent to roxie or other service implementations
out of band.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeck commented:
@mayx please review
@timothyklemm please review, and test if possible.

HPCCSmoketest commented:
Process of PR-11902, label: esdlTransformRoot starts now. Estimated completion time is ~ 0.5 hour(s)

afishbeck commented:
@timothyklemm to help with testing one option is using /opt/HPCCSystems/examples/EsdlExample. 

Compare running as is, and then changing esdl_binding.xml to reflect the differences here:

```
    <Method name="RoxieEchoPersonInfo" querytype="roxie" url="http://localhost:9876/roxie" queryname="RoxieEchoPersonInfo">
         <xsdl:CustomRequestTransform target="soap:Body/RoxieEchoPersonInfo">
            <xsdl:choose>
               <xsdl:when test="$clientversion=1.9">
                  <xsdl:SetValue target="vertest"  value="'v1.9'"/>
                  <xsdl:SetValue target="RoxieEchoPersonInfoRequest/Row/Name/First"  value="'v1.9'"/>
               </xsdl:when>
               <xsdl:otherwise>
                  <xsdl:SetValue target="vertest"  value="concat('v', $clientversion)"/>
                  <xsdl:SetValue target="RoxieEchoPersonInfoRequest/Row/Name/Last" value="concat('v', $clientversion)"/>
               </xsdl:otherwise>
            </xsdl:choose>
         </xsdl:CustomRequestTransform>
   </Method>
</Methods>
```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: afc69d7220042826bfcc996bacfae49d8c36fdcc
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   12 sec (00:00:12) |  188 sec (00:03:08) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1655 sec (00:27:35) |   19 sec (00:00:19) | 1894 sec (00:31:34) |

timothyklemm commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -357,12 +358,16 @@ void CEsdlCustomTransform::processTransform(IEspContext * context, StringBuffer
+        Owned<IPropertyTree> theroot = createPTreeFromXMLString(request.str());
+        StringBuffer xpath = m_target.str();
+        if (!xpath.length())
+            xpath.setf("soap:Body/%s/%s", mthdef.queryMethodName(), mthdef.queryRequestType()); //This default gives us backward compatibility with only being able to write to the actual request
```
Current roxie requests use an xpath like `soap:Body/MyRoxieService/MyESPMethodRequest`. This looks like it will generate `soap:Body/MyESPMethod/MyESPMethodRequest`. Will this work when MyRoxieService and MyESPMethod do not have the same value?

afishbeck replied:
Good point that may have just worked for my deployment.  It can always be over riden, but let me try to find where I can get the roxie service name from.

timothyklemm commented:
Another question. Is a reference outside of soap:Body ever valid, or could soap:Body be assumed for all targets, not just the default?

afishbeck commented:
@timothyklemm I didn't want to assume that soap:Body or whatever would always be a starting point.  For one thing someone may want to do something for SOAP compatibility issues... if calling a third part, or other service implementations (say for example JAVA plugins) may end up very different.  So I'll error on the side of complete flexibility.

afishbeck committed:
HPCC-20583 Allow DESDL application transforms to affect entire request

Some information can be sent to roxie or other service implementations
out of band.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@timothyklemm updated based on your comments.

Note that I changed the target path declaration to support variables:

<xsdl:CustomRequestTransform target="soap:Body/{$query}">

And fixed the default path to be query and request.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9cfd4036e9898dc57d1ecbfffa6631eb7c4a4a27
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  189 sec (00:03:09) |    0 sec (00:00:00) |    1 sec (00:00:01) |   18 sec (00:00:18) | 1456 sec (00:24:16) |   19 sec (00:00:19) | 1717 sec (00:28:37) |

afishbeck commented:
@timothyklemm please review.

timothyklemm reviewed (approved):
Changes look good. Haven't had a chance to try it out.

mayx reviewed (approved):
Looks good.

afishbeck commented:
@timothyklemm I agree with your comments, but think those can be covered by a future change and don't effect the first release.  The relevent service is very small and we can use method level transforms for now.  We can get the requested changes in soon after, and included in 7.2.

afishbeck commented:
@richardkchapman can you please merge?
